.github/workflows/general.yml

```diff
@@ -17,50 +19,86 @@ env:
     services:
+      # Label used to access the service container
       postgres:
+        # Docker Hub image 
         image: postgres:14
+        # Environment variables scoped only for the `postgres` element
         env:
           POSTGRES_USER: postgres
           POSTGRES_PASSWORD: password
           POSTGRES_DB: postgres
+        # When you map ports using the ports keyword, GitHub uses the --publish command to publish the container’s ports to the Docker host
+        # Opens tcp port 5432 on the host and service container
         ports:
           - 5432:5432
     steps:
-      - uses: actions/checkout@v3
-      - uses: dtolnay/rust-toolchain@stable
-      - uses: Swatinem/rust-cache@v2
+        # Downloads a copy of the code in your repository before running CI tests
+      - name: Check out repository code
+        # The uses keyword specifies that this step will run v3 of the actions/checkout action. 
+        # This is an action that checks out your repository onto the runner, allowing you to run scripts or other actions against your code (such as build and test tools).
+        # You should use the checkout action any time your workflow will run against the repository's code.
+        uses: actions/checkout@v3
+
+       # This GitHub Action installs a Rust toolchain using rustup. It is designed for one-line concise usage and good defaults.
+      - name: Install the Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+
+      # A GitHub Action that implements smart caching for rust/cargo projects with sensible defaults.
+      - name: Rust Cache Action
+        uses: Swatinem/rust-cache@v2
         with:
+          # An additional cache key that is added alongside the automatic `job`-based cache key and can be used to further differentiate jobs. default: empty
           key: sqlx-${{ env.SQLX_VERSION }}
+
       - name: Install sqlx-cli
         run:
             cargo install sqlx-cli
             --version=${{ env.SQLX_VERSION }}
             --features ${{ env.SQLX_FEATURES }}
             --no-default-features
             --locked
+            # The --locked flag can be used to force Cargo to use the packaged Cargo.lock file if it is available. 
+            # This may be useful for ensuring reproducible builds, to use the exact same set of dependencies that were available when the package was published. 
+            # It may also be useful if a newer version of a dependency is published that no longer builds on your system, or has other problems
+
+      - name: Install postgresql-client
+        run: sudo apt-get update && sudo apt-get install postgresql-client -y
+
       - name: Migrate database
-        run: |
-          sudo apt-get install libpq-dev -y
-          SKIP_DOCKER=true ./scripts/init_db.sh
+        run: SKIP_DOCKER=true ./scripts/init_db.sh
+
+      - name: Check sqlx-data.json is up-to-date
+        run: cargo sqlx prepare --check -- --bin zero2prod
```

Comment:
This is premature at this stage, we don't use the offline feature until chapter 5!